packages/flutter/lib/src/rendering/sliver_grid.dart

```diff
@@ -0,0 +1,400 @@
+      scrollOffset: (index ~/ crossAxisCount) * mainAxisStride,
+      crossAxisOffset: (index % crossAxisCount) * crossAxisStrid,
+      mainAxisExtent: childMainAxisExtent,
+      crossAxisExtent: childCrossAxisExtent,
+    );
+  }
+
+  @override
+  double estimateMaxScrollOffset(SliverConstraints constraints, int childCount) {
+    if (childCount == null)
+      return null;
+    final double crossAxisExtent = constraints.crossAxisExtent;
+    final int crossAxisCount = _getCrossAxisCount(crossAxisExtent);
+    final int mainAxisCount = ((childCount - 1) / crossAxisCount).floor() + 1;
+    return _getMainAxisStride(crossAxisExtent, crossAxisCount) * mainAxisCount - mainAxisSpacing;
+  }
+
+  @override
+  bool shouldRelayout(SliverGridDelegateWithMaxChildCrossAxisExtent oldDelegate) {
+    return oldDelegate.maxCrossAxisExtent != maxCrossAxisExtent
+        || oldDelegate.mainAxisSpacing != mainAxisSpacing
+        || oldDelegate.crossAxisSpacing != crossAxisSpacing
+        || oldDelegate.childAspectRatio != childAspectRatio;
+  }
+}
+
+class RenderSliverGrid extends RenderSliverMultiBoxAdaptor {
+  RenderSliverGrid({
+    @required RenderSliverBoxChildManager childManager,
+    @required SliverGridDelegate gridDelegate,
+  }) : _gridDelegate = gridDelegate,
+       super(childManager: childManager) {
+    gridDelegate != null;
+  }
+
+  @override
+  void setupParentData(RenderObject child) {
+    if (child.parentData is! SliverGridParentData)
+      child.parentData = new SliverGridParentData();
+  }
+
+  SliverGridDelegate get gridDelegate => _gridDelegate;
+  SliverGridDelegate _gridDelegate;
+
+  set gridDelegate(SliverGridDelegate newDelegate) {
+    assert(newDelegate != null);
+    if (_gridDelegate == newDelegate)
+      return;
+    if (newDelegate.runtimeType != _gridDelegate.runtimeType ||
+        newDelegate.shouldRelayout(_gridDelegate))
+      markNeedsLayout();
+    _gridDelegate = newDelegate;
+  }
+
+  @override
+  double childCrossAxisPosition(RenderBox child) {
+    final SliverGridParentData childParentData = child.parentData;
+    return childParentData.crossAxisOffset;
+  }
+
+  @override
+  void performLayout() {
+    assert(childManager.debugAssertChildListLocked());
+    final double scrollOffset = constraints.scrollOffset;
+    assert(scrollOffset >= 0.0);
+    final double remainingPaintExtent = constraints.remainingPaintExtent;
+    assert(remainingPaintExtent >= 0.0);
+    final double targetEndScrollOffset = scrollOffset + remainingPaintExtent;
+
+    final int firstIndex = _gridDelegate.getMinChildIndexForScrollOffset(
+        constraints, scrollOffset);
+    final int targetLastIndex = _gridDelegate.getMaxChildIndexForScrollOffset(
+        constraints, targetEndScrollOffset);
+
+    if (firstChild != null) {
+      final int oldFirstIndex = indexOf(firstChild);
+      final int oldLastIndex = indexOf(lastChild);
+      final int leadingGarbage = (firstIndex - oldFirstIndex).clamp(
```

Comment:
i'd just put this on one line. I don't think wrapping here is helping readability.